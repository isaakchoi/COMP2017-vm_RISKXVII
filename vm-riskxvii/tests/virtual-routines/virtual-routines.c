// Isaak Choi
// 520488399
// icho6322

// Constants
#define VR_WRITE_CHAR_ADDR       (0x0800) // Console Write Character
#define VR_WRITE_INT_ADDR        (0x0804) // Console Write Signed Integer
#define VR_WRITE_UINT_ADDR       (0x0808) // Console Write Unsigned Integer
#define VR_HALT_ADDR             (0x080C) // Halt
#define VR_READ_CHAR_ADDR        (0x0812) // Console Read Character
#define VR_READ_INT_ADDR         (0x0816) // Console Read Signed Integer
#define VR_DUMP_PC_ADDR          (0x0820) // Dump PC
#define VR_DUMP_REG_ADDR         (0x0824) // Dump Register Banks
#define VR_DUMP_MEM_WORD_ADDR    (0x0828) // Dump Memory Word
#define VR_HEAP_BANK_MALLOC_ADDR (0x0830) // Heap Bank - Malloc
#define VR_HEAP_BANK_FREE_ADDR   (0x0834) // Heap Bank - Free


////////////////////////////////////////////////////////////


// Get the address of the malloced memory
int get_malloc_addr(){
    int result;
    asm volatile("addi %[res], t3, 0"
        : [res]"=r"(result));
    return result;
}

// Print string to srdout
void print_string(char *str) {
    while (*str) *((char *)VR_WRITE_CHAR_ADDR) = *(str++);
}


////////////////////////////////////////////////////////////


// Read int - DONE
int get_int() {
    return *((int*)VR_READ_INT_ADDR);
}

// Read char - DONE
char get_char() {
    return *((int*)VR_READ_CHAR_ADDR);
}

// Write char - DONE
void write_char(char c) {
    *(int*)VR_WRITE_CHAR_ADDR = c; 
}

// Write int - DONE
void write_int(int n) {
    *(int*)VR_WRITE_INT_ADDR = n;
}

// Write uint - DONE
void write_uint(unsigned int n) {
    *(int*)VR_WRITE_UINT_ADDR = n;
}

// Halt - DONE
void halt() {
    *(int*)VR_HALT_ADDR = 1;
}

// Dump pc - DONE
void dump_pc() {
    *(int*)VR_DUMP_PC_ADDR = 1;
}

// Dump registers - DONE
void dump_reg() {
    *(int*)VR_DUMP_REG_ADDR = 1;
}

// Dump mem word - DONE
void dump_word(int addr) {
    *(int*)VR_DUMP_MEM_WORD_ADDR = addr;
}

// malloc - DONE
void vr_malloc(int size) {
    *(int*)VR_HEAP_BANK_MALLOC_ADDR = size;
}

// free - DONE
void vr_free(int addr) {
    *(int*)VR_HEAP_BANK_FREE_ADDR = addr;
}


// Main
int main(){


    print_string("Enter single char: ");
    char c = get_char();

    print_string("\nEnter single int: ");
    int n = get_int();

    print_string("\nGiven char: ");
    write_char(c);

    print_string("\nGiven int: ");
    write_int(n);

    print_string("\nGiven int as uint: ");
    write_uint(n);

    print_string("\nDumping pc:\n");
    dump_pc();

    print_string("\nDumping registers:\n");
    dump_reg();

    print_string("\nDumping memory word at address of given int:\n");
    dump_word((int)&n);

    print_string("\nmalloc'ing 128 individual banks ...\n");
    vr_malloc(1);
    int first_malloc_addr = get_malloc_addr();
    for (int i = 0; i < 127; i++) {
        vr_malloc(1);
    }
    print_string("DONE\n");

    print_string("free'ing all malloc banks ...\n");
    for (int i = 0; i < 128; i++) {
        #define BANK_SIZE_BYTES (64)
        vr_free(first_malloc_addr);
        first_malloc_addr += BANK_SIZE_BYTES;
    }
    print_string("DONE\n");

    print_string("HALTING ...\n");
    halt();
    print_string("This printed after halt >:(\n");

    // Return
    return 0;
}

