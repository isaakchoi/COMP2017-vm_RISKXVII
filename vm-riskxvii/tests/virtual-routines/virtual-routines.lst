
virtual-routines/virtual-routines:     file format elf32-littleriscv


Disassembly of section .text:

00000000 <_start>:
   0:	7ff00113          	li	sp,2047
   4:	280000ef          	jal	ra,284 <main>
   8:	000017b7          	lui	a5,0x1
   c:	80078623          	sb	zero,-2036(a5) # 80c <ADDR_END+0xc>

00000010 <get_malloc_addr>:
  10:	fe010113          	addi	sp,sp,-32
  14:	00812e23          	sw	s0,28(sp)
  18:	02010413          	addi	s0,sp,32
  1c:	000e0793          	mv	a5,t3
  20:	fef42623          	sw	a5,-20(s0)
  24:	fec42783          	lw	a5,-20(s0)
  28:	00078513          	mv	a0,a5
  2c:	01c12403          	lw	s0,28(sp)
  30:	02010113          	addi	sp,sp,32
  34:	00008067          	ret

00000038 <print_string>:
  38:	fe010113          	addi	sp,sp,-32
  3c:	00812e23          	sw	s0,28(sp)
  40:	02010413          	addi	s0,sp,32
  44:	fea42623          	sw	a0,-20(s0)
  48:	0200006f          	j	68 <print_string+0x30>
  4c:	fec42783          	lw	a5,-20(s0)
  50:	00178713          	addi	a4,a5,1
  54:	fee42623          	sw	a4,-20(s0)
  58:	00001737          	lui	a4,0x1
  5c:	80070713          	addi	a4,a4,-2048 # 800 <ADDR_END>
  60:	0007c783          	lbu	a5,0(a5)
  64:	00f70023          	sb	a5,0(a4)
  68:	fec42783          	lw	a5,-20(s0)
  6c:	0007c783          	lbu	a5,0(a5)
  70:	fc079ee3          	bnez	a5,4c <print_string+0x14>
  74:	00000013          	nop
  78:	00000013          	nop
  7c:	01c12403          	lw	s0,28(sp)
  80:	02010113          	addi	sp,sp,32
  84:	00008067          	ret

00000088 <get_int>:
  88:	ff010113          	addi	sp,sp,-16
  8c:	00812623          	sw	s0,12(sp)
  90:	01010413          	addi	s0,sp,16
  94:	000017b7          	lui	a5,0x1
  98:	81678793          	addi	a5,a5,-2026 # 816 <ADDR_END+0x16>
  9c:	0007a783          	lw	a5,0(a5)
  a0:	00078513          	mv	a0,a5
  a4:	00c12403          	lw	s0,12(sp)
  a8:	01010113          	addi	sp,sp,16
  ac:	00008067          	ret

000000b0 <get_char>:
  b0:	ff010113          	addi	sp,sp,-16
  b4:	00812623          	sw	s0,12(sp)
  b8:	01010413          	addi	s0,sp,16
  bc:	000017b7          	lui	a5,0x1
  c0:	81278793          	addi	a5,a5,-2030 # 812 <ADDR_END+0x12>
  c4:	0007a783          	lw	a5,0(a5)
  c8:	0ff7f793          	zext.b	a5,a5
  cc:	00078513          	mv	a0,a5
  d0:	00c12403          	lw	s0,12(sp)
  d4:	01010113          	addi	sp,sp,16
  d8:	00008067          	ret

000000dc <write_char>:
  dc:	fe010113          	addi	sp,sp,-32
  e0:	00812e23          	sw	s0,28(sp)
  e4:	02010413          	addi	s0,sp,32
  e8:	00050793          	mv	a5,a0
  ec:	fef407a3          	sb	a5,-17(s0)
  f0:	000017b7          	lui	a5,0x1
  f4:	80078793          	addi	a5,a5,-2048 # 800 <ADDR_END>
  f8:	fef44703          	lbu	a4,-17(s0)
  fc:	00e7a023          	sw	a4,0(a5)
 100:	00000013          	nop
 104:	01c12403          	lw	s0,28(sp)
 108:	02010113          	addi	sp,sp,32
 10c:	00008067          	ret

00000110 <write_int>:
 110:	fe010113          	addi	sp,sp,-32
 114:	00812e23          	sw	s0,28(sp)
 118:	02010413          	addi	s0,sp,32
 11c:	fea42623          	sw	a0,-20(s0)
 120:	000017b7          	lui	a5,0x1
 124:	80478793          	addi	a5,a5,-2044 # 804 <ADDR_END+0x4>
 128:	fec42703          	lw	a4,-20(s0)
 12c:	00e7a023          	sw	a4,0(a5)
 130:	00000013          	nop
 134:	01c12403          	lw	s0,28(sp)
 138:	02010113          	addi	sp,sp,32
 13c:	00008067          	ret

00000140 <write_uint>:
 140:	fe010113          	addi	sp,sp,-32
 144:	00812e23          	sw	s0,28(sp)
 148:	02010413          	addi	s0,sp,32
 14c:	fea42623          	sw	a0,-20(s0)
 150:	000017b7          	lui	a5,0x1
 154:	80878793          	addi	a5,a5,-2040 # 808 <ADDR_END+0x8>
 158:	fec42703          	lw	a4,-20(s0)
 15c:	00e7a023          	sw	a4,0(a5)
 160:	00000013          	nop
 164:	01c12403          	lw	s0,28(sp)
 168:	02010113          	addi	sp,sp,32
 16c:	00008067          	ret

00000170 <halt>:
 170:	ff010113          	addi	sp,sp,-16
 174:	00812623          	sw	s0,12(sp)
 178:	01010413          	addi	s0,sp,16
 17c:	000017b7          	lui	a5,0x1
 180:	80c78793          	addi	a5,a5,-2036 # 80c <ADDR_END+0xc>
 184:	00100713          	li	a4,1
 188:	00e7a023          	sw	a4,0(a5)
 18c:	00000013          	nop
 190:	00c12403          	lw	s0,12(sp)
 194:	01010113          	addi	sp,sp,16
 198:	00008067          	ret

0000019c <dump_pc>:
 19c:	ff010113          	addi	sp,sp,-16
 1a0:	00812623          	sw	s0,12(sp)
 1a4:	01010413          	addi	s0,sp,16
 1a8:	000017b7          	lui	a5,0x1
 1ac:	82078793          	addi	a5,a5,-2016 # 820 <ADDR_END+0x20>
 1b0:	00100713          	li	a4,1
 1b4:	00e7a023          	sw	a4,0(a5)
 1b8:	00000013          	nop
 1bc:	00c12403          	lw	s0,12(sp)
 1c0:	01010113          	addi	sp,sp,16
 1c4:	00008067          	ret

000001c8 <dump_reg>:
 1c8:	ff010113          	addi	sp,sp,-16
 1cc:	00812623          	sw	s0,12(sp)
 1d0:	01010413          	addi	s0,sp,16
 1d4:	000017b7          	lui	a5,0x1
 1d8:	82478793          	addi	a5,a5,-2012 # 824 <ADDR_END+0x24>
 1dc:	00100713          	li	a4,1
 1e0:	00e7a023          	sw	a4,0(a5)
 1e4:	00000013          	nop
 1e8:	00c12403          	lw	s0,12(sp)
 1ec:	01010113          	addi	sp,sp,16
 1f0:	00008067          	ret

000001f4 <dump_word>:
 1f4:	fe010113          	addi	sp,sp,-32
 1f8:	00812e23          	sw	s0,28(sp)
 1fc:	02010413          	addi	s0,sp,32
 200:	fea42623          	sw	a0,-20(s0)
 204:	000017b7          	lui	a5,0x1
 208:	82878793          	addi	a5,a5,-2008 # 828 <ADDR_END+0x28>
 20c:	fec42703          	lw	a4,-20(s0)
 210:	00e7a023          	sw	a4,0(a5)
 214:	00000013          	nop
 218:	01c12403          	lw	s0,28(sp)
 21c:	02010113          	addi	sp,sp,32
 220:	00008067          	ret

00000224 <vr_malloc>:
 224:	fe010113          	addi	sp,sp,-32
 228:	00812e23          	sw	s0,28(sp)
 22c:	02010413          	addi	s0,sp,32
 230:	fea42623          	sw	a0,-20(s0)
 234:	000017b7          	lui	a5,0x1
 238:	83078793          	addi	a5,a5,-2000 # 830 <ADDR_END+0x30>
 23c:	fec42703          	lw	a4,-20(s0)
 240:	00e7a023          	sw	a4,0(a5)
 244:	00000013          	nop
 248:	01c12403          	lw	s0,28(sp)
 24c:	02010113          	addi	sp,sp,32
 250:	00008067          	ret

00000254 <vr_free>:
 254:	fe010113          	addi	sp,sp,-32
 258:	00812e23          	sw	s0,28(sp)
 25c:	02010413          	addi	s0,sp,32
 260:	fea42623          	sw	a0,-20(s0)
 264:	000017b7          	lui	a5,0x1
 268:	83478793          	addi	a5,a5,-1996 # 834 <ADDR_END+0x34>
 26c:	fec42703          	lw	a4,-20(s0)
 270:	00e7a023          	sw	a4,0(a5)
 274:	00000013          	nop
 278:	01c12403          	lw	s0,28(sp)
 27c:	02010113          	addi	sp,sp,32
 280:	00008067          	ret

00000284 <main>:
 284:	fd010113          	addi	sp,sp,-48
 288:	02112623          	sw	ra,44(sp)
 28c:	02812423          	sw	s0,40(sp)
 290:	03010413          	addi	s0,sp,48
 294:	40000513          	li	a0,1024
 298:	da1ff0ef          	jal	ra,38 <print_string>
 29c:	e15ff0ef          	jal	ra,b0 <get_char>
 2a0:	00050793          	mv	a5,a0
 2a4:	fef401a3          	sb	a5,-29(s0)
 2a8:	41400513          	li	a0,1044
 2ac:	d8dff0ef          	jal	ra,38 <print_string>
 2b0:	dd9ff0ef          	jal	ra,88 <get_int>
 2b4:	00050793          	mv	a5,a0
 2b8:	fcf42e23          	sw	a5,-36(s0)
 2bc:	42800513          	li	a0,1064
 2c0:	d79ff0ef          	jal	ra,38 <print_string>
 2c4:	fe344783          	lbu	a5,-29(s0)
 2c8:	00078513          	mv	a0,a5
 2cc:	e11ff0ef          	jal	ra,dc <write_char>
 2d0:	43800513          	li	a0,1080
 2d4:	d65ff0ef          	jal	ra,38 <print_string>
 2d8:	fdc42783          	lw	a5,-36(s0)
 2dc:	00078513          	mv	a0,a5
 2e0:	e31ff0ef          	jal	ra,110 <write_int>
 2e4:	44800513          	li	a0,1096
 2e8:	d51ff0ef          	jal	ra,38 <print_string>
 2ec:	fdc42783          	lw	a5,-36(s0)
 2f0:	00078513          	mv	a0,a5
 2f4:	e4dff0ef          	jal	ra,140 <write_uint>
 2f8:	46000513          	li	a0,1120
 2fc:	d3dff0ef          	jal	ra,38 <print_string>
 300:	e9dff0ef          	jal	ra,19c <dump_pc>
 304:	47000513          	li	a0,1136
 308:	d31ff0ef          	jal	ra,38 <print_string>
 30c:	ebdff0ef          	jal	ra,1c8 <dump_reg>
 310:	48800513          	li	a0,1160
 314:	d25ff0ef          	jal	ra,38 <print_string>
 318:	fdc40793          	addi	a5,s0,-36
 31c:	00078513          	mv	a0,a5
 320:	ed5ff0ef          	jal	ra,1f4 <dump_word>
 324:	4b800513          	li	a0,1208
 328:	d11ff0ef          	jal	ra,38 <print_string>
 32c:	00100513          	li	a0,1
 330:	ef5ff0ef          	jal	ra,224 <vr_malloc>
 334:	cddff0ef          	jal	ra,10 <get_malloc_addr>
 338:	fea42623          	sw	a0,-20(s0)
 33c:	fe042423          	sw	zero,-24(s0)
 340:	0180006f          	j	358 <main+0xd4>
 344:	00100513          	li	a0,1
 348:	eddff0ef          	jal	ra,224 <vr_malloc>
 34c:	fe842783          	lw	a5,-24(s0)
 350:	00178793          	addi	a5,a5,1
 354:	fef42423          	sw	a5,-24(s0)
 358:	fe842703          	lw	a4,-24(s0)
 35c:	07e00793          	li	a5,126
 360:	fee7d2e3          	bge	a5,a4,344 <main+0xc0>
 364:	4e000513          	li	a0,1248
 368:	cd1ff0ef          	jal	ra,38 <print_string>
 36c:	4e800513          	li	a0,1256
 370:	cc9ff0ef          	jal	ra,38 <print_string>
 374:	fe042223          	sw	zero,-28(s0)
 378:	0240006f          	j	39c <main+0x118>
 37c:	fec42503          	lw	a0,-20(s0)
 380:	ed5ff0ef          	jal	ra,254 <vr_free>
 384:	fec42783          	lw	a5,-20(s0)
 388:	04078793          	addi	a5,a5,64
 38c:	fef42623          	sw	a5,-20(s0)
 390:	fe442783          	lw	a5,-28(s0)
 394:	00178793          	addi	a5,a5,1
 398:	fef42223          	sw	a5,-28(s0)
 39c:	fe442703          	lw	a4,-28(s0)
 3a0:	07f00793          	li	a5,127
 3a4:	fce7dce3          	bge	a5,a4,37c <main+0xf8>
 3a8:	4e000513          	li	a0,1248
 3ac:	c8dff0ef          	jal	ra,38 <print_string>
 3b0:	50800513          	li	a0,1288
 3b4:	c85ff0ef          	jal	ra,38 <print_string>
 3b8:	db9ff0ef          	jal	ra,170 <halt>
 3bc:	51800513          	li	a0,1304
 3c0:	c79ff0ef          	jal	ra,38 <print_string>
 3c4:	00000793          	li	a5,0
 3c8:	00078513          	mv	a0,a5
 3cc:	02c12083          	lw	ra,44(sp)
 3d0:	02812403          	lw	s0,40(sp)
 3d4:	03010113          	addi	sp,sp,48
 3d8:	00008067          	ret
