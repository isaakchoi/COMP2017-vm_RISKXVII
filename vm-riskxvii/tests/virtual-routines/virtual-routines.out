Enter single char: 
Enter single int: 
Given char: X
Given int: 8
Given int as uint: 8
Dumping pc:
1b4
Dumping registers:
PC = 0x000001e0;
R[0] = 0x00000000;
R[1] = 0x00000310;
R[2] = 0x000007bf;
R[3] = 0x00000000;
R[4] = 0x00000000;
R[5] = 0x00000000;
R[6] = 0x00000000;
R[7] = 0x00000000;
R[8] = 0x000007cf;
R[9] = 0x00000000;
R[10] = 0x00000470;
R[11] = 0x00000000;
R[12] = 0x00000000;
R[13] = 0x00000000;
R[14] = 0x00000001;
R[15] = 0x00000824;
R[16] = 0x00000000;
R[17] = 0x00000000;
R[18] = 0x00000000;
R[19] = 0x00000000;
R[20] = 0x00000000;
R[21] = 0x00000000;
R[22] = 0x00000000;
R[23] = 0x00000000;
R[24] = 0x00000000;
R[25] = 0x00000000;
R[26] = 0x00000000;
R[27] = 0x00000000;
R[28] = 0x00000000;
R[29] = 0x00000000;
R[30] = 0x00000000;
R[31] = 0x00000000;

Dumping memory word at address of given int:
8
malloc'ing 128 individual banks ...
DONE
free'ing all malloc banks ...
DONE
HALTING ...
CPU Halt Requested
